xrdp/xrdp_egfx.c

```diff
@@ -0,0 +1,1021 @@
+    bytes = bitmap_data_length + 8192;
+    bytes += 5 * (bitmap_data_length / 0xFFFF);
+    init_stream(s, bytes);
+    /* RDP_SEGMENTED_DATA */
+    out_uint8(s, 0xE1); /* descriptor = MULTIPART */
+    hold_segment_count = s->p;
+    out_uint8s(s, 2); /* segmentCount set later */
+    out_uint32_le(s, 21 + bitmap_data_length); /* uncompressedSize */
+    /* RDP_DATA_SEGMENT */
+    out_uint32_le(s, 1 + 21); /* segmentArray size */
+    /* RDP8_BULK_ENCODED_DATA */
+    out_uint8(s, 0x04); /* header = PACKET_COMPR_TYPE_RDP8 */
+    /* RDPGFX_HEADER */
+    out_uint16_le(s, XR_RDPGFX_CMDID_WIRETOSURFACE_2); /* cmdId */
+    out_uint16_le(s, 0); /* flags = 0 */
+    out_uint32_le(s, 21 + bitmap_data_length); /* pduLength */
+    out_uint16_le(s, surface_id);
+    out_uint16_le(s, codec_id);
+    out_uint32_le(s, codec_context_id);
+    out_uint8(s, pixel_format);
+    out_uint32_le(s, bitmap_data_length);
+    segment_count = 1;
+    index = 0;
+    bitmap_data8 = (char *) bitmap_data;
+    while (index < bitmap_data_length)
+    {
+        segment_size = bitmap_data_length - index;
+        if (segment_size > USHRT_MAX)
+        {
+            segment_size = USHRT_MAX;
+        }
+        /* RDP_DATA_SEGMENT */
+        out_uint32_le(s, 1 + segment_size); /* segmentArray size */
+        /* RDP8_BULK_ENCODED_DATA */
+        out_uint8(s, 0x04); /* header = PACKET_COMPR_TYPE_RDP8 */
+        out_uint8a(s, bitmap_data8 + index, segment_size);
+        LOG(LOG_LEVEL_DEBUG, "  segment index %d"
+            " segment_size %d", segment_count, segment_size);
+        index += segment_size;
+        segment_count++;
+    }
+    s_mark_end(s);
+    bytes = (int) (s->end - s->data);
+    /* segment_count */
+    s->p = hold_segment_count;
+    out_uint16_le(s, segment_count);
+    error = xrdp_egfx_send_data(egfx, s->data, bytes);
+    LOG(LOG_LEVEL_DEBUG, "xrdp_egfx_send_wire_to_surface2:"
+        " xrdp_egfx_send_data error %d segment_count %d", error, segment_count);
+    free_stream(s);
+    return error;
+}
+
+/******************************************************************************/
+int
+xrdp_egfx_send_reset_graphics(struct xrdp_egfx *egfx, int width, int height,
+                              int monitor_count, struct monitor_info *mi)
+{
+    int error;
+    int bytes;
+    int index;
+    struct stream *s;
+
+    LOG(LOG_LEVEL_INFO, "xrdp_egfx_send_reset_graphics:");
+    if (monitor_count > 16)
+    {
+        return 1;
+    }
+    make_stream(s);
+    init_stream(s, 8192);
+    /* RDP_SEGMENTED_DATA */
+    out_uint8(s, 0xE0); /* descriptor = SINGLE */
+    /* RDP8_BULK_ENCODED_DATA */
+    out_uint8(s, 0x04); /* header = PACKET_COMPR_TYPE_RDP8 */
+    /* RDPGFX_HEADER */
+    out_uint16_le(s, XR_RDPGFX_CMDID_RESETGRAPHICS); /* cmdId */
+    out_uint16_le(s, 0); /* flags = 0 */
+    out_uint32_le(s, 340);
+    out_uint32_le(s, width);
+    out_uint32_le(s, height);
+    out_uint32_le(s, monitor_count == 0 ? 1 : monitor_count);
```

Comment:
Since I am not an xrdp member, its your choice. But personally I wouldn't apply workarounds for the own code here.

Reply:
This is a workaround for Microsoft's Mac OS Client. We can't ever send a monitorcount of 0, even though internally sometimes it's set that way. So, from Microsoft's perspective, this is important to always be one.

What would you recommend I do instead?

Reply:
> We can't ever send a monitorcount of 0

Well, it's RDP, it's about graphical remote desktop. You always have at least one monitor. Sending a monitor config here with 0 monitors is a bug.
If you know, that the monitor config changed, but not know yet the actual monitor layout, then defer sending this PDU until you know the new layout.
Same for the subsequent PDUs, which provide the graphics content (wire_to_surface, etc.).
There is no need to send this PDU until the screen changes or the layout changes are done.

The question would also be: If you have 0 monitors, what size for the graphics output buffer do you use in that situation?
The graphics output buffer only contains user visible surfaces, i.e. surfaces that are mapped to an output or a window (map_surface_to_[output,window].
You can have offscreen surfaces, that are not mapped, that are actually larger than the graphics output buffer itself, in case that is relevant for you.

Reply:
I agree that 0 monitors is a bug in XRDP, but it's pretty baked in at this point and probably out of scope for this PR to fix it.

In the past "0 monitors with a session_width and a session_height" just meant "multi-mon is not enabled and we have one single large surface." It didn't actually matter until GFX integration, which cares about monitor numbering more than RLE or RemoteFX appeared to.

That is indeed wrong. I'll write up an issue about what's wrong here and that we should fix it (and hopefully if I have time, what needs to be fixed), then reference it in a TODO.

Sound good?

Reply:
> I agree that 0 monitors is a bug in XRDP, but it's pretty baked in at this point and probably out of scope for this PR to fix it.

Sure, `xrdp_egfx_send_reset_graphics()` is also not used yet.

Personally, I would, in case that bug is unavoidable, apply the workaround at the earliest point possible (i.e. in future commits, where `xrdp_egfx_send_reset_graphics` is actually used). This way, the area of code, where that bug exists would be reduced.

> Sound good?

Ultimately, it is up to you how you handle that, because I am not part of the xrdp project. I can only point out wrong or fishy things here.

> I'll write up an issue about what's wrong here and that we should fix it (and hopefully if I have time, what needs to be fixed), then reference it in a TODO.

Creating an issue sounds sane to avoid, that the issue is lost, while the bug still exists.